_api-reference/cluster-api/cluster-settings.md

```diff
@@ -102,6 +102,7 @@ The following request field parameters are compatible with the cluster API.
 | cluster.routing.allocation.include.<attribute> | Enum | Allocates shards to a node whose `attribute` has at least one of the included comma-separated values. |
 | cluster.routing.allocation.require.<attribute> | Enum | Only allocates shards to a node whose `attribute` has all of the included comma-separated values. |
 | cluster.routing.allocation.exclude.<attribute> | Enum | Does not allocate shards to a node whose `attribute` has any of the included comma-separated values. The cluster allocation settings support the following built-in attributes: <br /> <br /> `_name` – Match nodes by node name. <br /> <br /> `_host_ip` – Match nodes by host IP address. <br /> <br /> `_publish_ip` – Match nodes by publish IP address. <br /> <br /> `_ip` – Match either `_host_ip` or `_publish_ip`. <br /> <br /> `_host` – Match nodes by hostname. <br /> <br /> `_id` – Match nodes by node ID. <br /> <br /> `_tier` – Match nodes by data tier role. |
+| cluster.routing.allocation.shard_movement_strategy | Enum |  Decides order in which to move shards from node when shards can not stay on node anymore. Supports the following strategies: <br /> <br /> `NO_PREFERENCE` - default behavior in which order of shard movement doesn't matter. <br /> <br /> `PRIMARY_FIRST` - primary shards are moved first. <br /> <br /> `REPLICA_FIRST` - replica shards are moved first.
```

Comment:
1. These settings were introduced to deal with specific scenarios. 

-  `PRIMARY_FIRST` prioritizes the primary of all shards first so the cluster does not become red if any excluded nodes (based on a cluster setting) were to go down before relocating other shards. 
-  `REPLICA_FIRST` prioritizes the replica of all shards first so the cluster does not become red in a mixed version cluster if primaries are relocated to higher version nodes and send segments on a higher OS version to replicas that they cannot read, subsequently leading to shard failure. 

Note that with this setting enabled performance of this change is a direct function of number of indices, shards, replicas, and nodes. The larger the indices, replicas, and distribution scale, the slower the allocation becomes. This should be used with care. This is true for both the above settings. 

2. No logic that determines shard movement order. It is random. 

We can add in some information from (1) regarding performance and use cases for the settings. I had omitted that since the table of request field parameters didn't seem the right place to add in all this information - but if it's preferred that we add it here I'm good with including some more. 

Let me know and I'll make the changes including your suggestion :)

Reply:
@Poojita-Raj Thanks for adding context here. I think including something about these two scenarios would be worthwhile. Have a look at these revisions to the setting descriptions and let me know if I misunderstood your explanations (I'm not familiar with cluster health and shard allocation. So this is highly possible):

`PRIMARY_FIRST` -- Primary shards are relocated first, before replica shards. This prioritization may help prevent a cluster's health from going red if nodes excluded _from shard relocation_ happened to go down during the process.

1. I'm not sure what the nodes are excluded from. Are they excluded from shard relocation? Which cluster setting are we talking about here? 
2. In the original, your mentioned "before relocating other shards". I'm not sure what these "other shards" are. Are these primary or replica shards? Are these the replica shards that would get second priority to the primaries in this configuration?

`REPLICA_FIRST` -- Replica shards are moved first, before primary shards. This prioritization may help prevent a cluster's health from going red when carrying out shard relocation in a cluster that includes nodes on different versions of OpenSearch. In this situation, primary shards relocated to later-version nodes could try to copy segment files to replica shards on an earlier version of OpenSearch, which would result in shard failure. Relocating replica shards first attempts to avoid this from happening in multi-version clusters.

1. Is this the idea here, to relocate the replica shards first and then relocate primary shards (including those on newer versions of OpenSearch) and remove the possibility of segment file read errors?

`NO_PREFERENCE` - The default behavior in which the order of shard relocation has no importance.

Let's include these descriptions in this order. It seems the purpose of adding this new setting is to give users an important choice to relocate shards one way or another. Let's present them with the options and then, last, tell them the default doesn't consider order.

Reply:
@cwillum Thanks for the suggestions! I'll change the order of the settings and include the descriptions for each setting. 

`PRIMARY_FIRST` - (1) When you exclude a group of nodes from a cluster, the shards on those nodes are moved to other nodes in the cluster. `cluster.routing.allocation.exclude.` is the setting used for this. 
(2) Yes, other shards refers to replica shards in this case that would get second priority. 

`REPLICA_FIRST` - (1) Yes, that's exactly it.